Source/DafnyTestGeneration/Auditor.cs

```diff
@@ -0,0 +1,373 @@
+using Type = Microsoft.Dafny.Type;
+
+namespace DafnyTestGeneration;
+
+// This AST pass is run before test generation to detect any potential issues and report them to the user
+public class Auditor {
+
+  private List<DafnyDiagnostic> diagnostics;
+  private IEnumerable<DafnyDiagnostic> Errors => diagnostics.Where(diagnostic => diagnostic.Level == ErrorLevel.Error);
+  private IEnumerable<DafnyDiagnostic> Warnings => diagnostics.Where(diagnostic => diagnostic.Level == ErrorLevel.Warning);
+  private readonly DafnyOptions options;
+  public bool NonZeroExitCode { get; private set; }
+
+  // Errors always lead to preemptive termination of test generation and cannot be bypassed
+  public const string NoTestEntryError = "NoTestEntryError";
+  public const string NoExternalModuleError = "NoExternalModuleError";
+  public const string MalformedAttributeError = "MalformedAttributeError";
+  public const string UnsupportedInputTypeError = "UnsupportedInputTypeError";
+
+  // Warnings lead to preemptive terminate of test generation unless --ignore-warnings flag is used
+  public const string InlinedMethodNotReachableWarning = "InlinedMethodNotReachableWarning";
+  public const string NotFullySupportedInputTypeWarning = "NotFullySupportedInputTypeWarning";
+  public const string NoWitnessWarning = "NoWitnessWarning";
+  public const string SmallTimeLimitWarning = "SmallTumeLimitWarning";
+
+  // List of all types that have been checked for being supported. Used to prevent infinite recursion
+  private HashSet<Type> typesConsidered = new();
+
+  public Auditor(DafnyOptions options) {
+    this.options = options;
+  }
+
+  /// <summary>
+  /// Check that test generation will be able to process the program.
+  /// Return false, if test generation should preemptively terminated,
+  /// i.e. if there are any errors or if there are warnings and the --ignore-warnings flag is not used
+  /// </summary>
+  public bool IsOk(string source, Uri uri) {
+    var errorReporter = new ConsoleErrorReporter(options);
+    var program = Utils.Parse(errorReporter, source, true, uri);
+    diagnostics = new();
+    if (errorReporter.HasErrors) {
+      NonZeroExitCode = true;
+      return false;
+    }
+    typesConsidered = new();
+    CheckIsWrappedInAModule(program);
+    CheckHasTestEntry(program);
+    CheckInlinedDeclarationsAreReachable(program);
+    CheckInlineAttributes(program);
+    CheckInputTypesAreSupported(program);
+    CheckVerificationTimeLimit(program);
+    PrintWarningsAndErrors(errorReporter, program);
+    NonZeroExitCode = Errors.Count() != 0;
+    return !Errors.Any() && (!Warnings.Any() || !options.WarningsAsErrors);
+  }
+
+  /// <summary>
+  /// Print all gathered warnings and errors
+  /// </summary>
+  private void PrintWarningsAndErrors(ErrorReporter errorReporter, Program program) {
+    if (Errors.Count() != 0) {
+      errorReporter.Message(MessageSource.TestGenerationAuditor, ErrorLevel.Error, "", program.StartToken,
```

Comment:
Is there a reason you can't use the more concise `Error` method instead of the `Message` method?

Reply:
No, I think I just wasn't aware that can use Error/Message - have changed the code to use them now